source/dataman/CLS/CLSDbUpgrade1Pt1.cpp

```diff
@@ -0,0 +1,128 @@
+#include "CLSDbUpgrade1Pt1.h"
+
+#include "util/AMErrorMonitor.h"
+
+CLSDbUpgrade1Pt1::CLSDbUpgrade1Pt1(CLSAppController::CLSBeamlineID targetFacilityId, const QString &databaseNameToUpgrade, QObject *parent)
+	: AMDbUpgrade(databaseNameToUpgrade, parent)
+{
+	targetFacilityId_ = targetFacilityId;
+}
+
+CLSDbUpgrade1Pt1::~CLSDbUpgrade1Pt1()
+{
+
+}
+
+QStringList CLSDbUpgrade1Pt1::upgradeFromTags() const
+{
+	return QStringList();
+}
+
+bool CLSDbUpgrade1Pt1::upgradeNecessary() const
+{
+	// run update when:
+	//
+	//   - the facilityId column doesn't exist for AMFacility or
+	//   - the total number of facilities doesn't match with the existing max values
+
+	bool hasFacilityId = databaseToUpgrade_->tableExists("AMFacility_table") && databaseToUpgrade_->columnExists("AMFacility_table", "facilityId");
+
+	QVariantList tableFacilityNames = databaseToUpgrade_->retrieve("AMFacility_table", "name");
+	bool matchFacilityCount = (tableFacilityNames.count() == CLSAppController::CLSMAXBeamlineId - 1);
```

Comment:
Yes. the design of AMFacility table works in that way. AMFacility table holds all the current beamline applications.